[Automation] native-image run fails for io.netty:netty-codec-memcache:4.2.1.Final

## native-image run on ubuntu-latest / JDK 25 / mode current-defaults

Failure kind: native-image run
OS: ubuntu-latest
JDK: 25
Native-image mode: current-defaults
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.2.1.Final" ./gradlew clean test -Pcoordinates="io.netty:netty-codec-memcache:4.1.79.Final"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25296069081/job/74155118831
Last 300 lines of the log (excerpt truncated to fit the GitHub issue size limit):
```console
work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/junit/junit/4.13.2,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/org.hamcrest/hamcrest-core/1.3,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/io.netty/netty-codec-memcache/4.1.79.Final,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/io.netty/netty-common/4.1.115.Final,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/io.netty/netty-buffer/4.1.80.Final,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/io.netty/netty-transport/4.1.115.Final,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/io.netty/netty-resolver/4.1.79.Final \
-H:Optimize@user+api=b \
-H:+ReportExceptionStackTraces@user \
-H:Features@user+api=org.graalvm.junit.platform.JUnitPlatformFeature \
'-H:Class@explicit main-class=org.graalvm.junit.platform.NativeImageJUnitLauncher' \
-H:ImageBuildID@driver=1b9a29fb-7725-0190-afda-dd753c1440fe \
'-H:Features@jar:file:///opt/hostedtoolcache/graalvm-jdk-25_linux-x64_bin/25.0.0/x64/graalvm-jdk-25.0.3+9.1/lib/svm/library-support.jar!/META-INF/native-image/com.oracle.svm/thirdparty/native-image.properties+api=com.oracle.svm.thirdparty.gson.GsonFeature' \
'-H:Features@jar:file:///opt/hostedtoolcache/graalvm-jdk-25_linux-x64_bin/25.0.0/x64/graalvm-jdk-25.0.3+9.1/lib/svm/library-support.jar!/META-INF/native-image/com.oracle.svm/polyglot/native-image.properties+api=com.oracle.svm.polyglot.groovy.GroovyIndyInterfaceFeature,com.oracle.svm.polyglot.scala.ScalaFeature'
]
========================================================================================================================
GraalVM Native Image: Generating 'io.netty.netty-codec-memcache_tests-tests' (executable)...
========================================================================================================================
For detailed information and explanations on the build output, visit:
https://github.com/oracle/graal/blob/master/docs/reference-manual/native-image/BuildOutput.md
------------------------------------------------------------------------------------------------------------------------
[1/8] Initializing...                                                                                    (6.5s @ 0.14GB)
 Java version: 25.0.3+9-LTS, vendor version: Oracle GraalVM 25.0.3+9.1
 Graal compiler: optimization level: b, target machine: x86-64-v3, PGO: off
 C compiler: gcc (linux, x86_64, 13.3.0)
 Garbage collector: Serial GC (max heap size: 80% of RAM)
 2 user-specific feature(s):
 - com.oracle.svm.thirdparty.gson.GsonFeature
 - org.graalvm.junit.platform.JUnitPlatformFeature
------------------------------------------------------------------------------------------------------------------------
 1 experimental option(s) unlocked:
 - '-H:BuildOutputJSONFile' (origin(s): command line)
------------------------------------------------------------------------------------------------------------------------
Build resources:
 - 12.67GB of memory (75.6% of system memory, $CI set to 'true')
 - 4 thread(s) (100.0% of 4 available processor(s), determined at start)
[junit-platform-native] Running in 'test listener' mode using files matching pattern [junit-platform-unique-ids*] found in folder [/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/test-results/test/testlist] and its subfolders.
[2/8] Performing analysis...  [******]                                                                  (37.6s @ 0.92GB)
    8,900 types,  10,124 fields, and  39,980 methods found reachable
    3,216 types,     803 fields, and   4,217 methods registered for reflection
       62 types,      68 fields, and      53 methods registered for JNI access
        0 downcalls and 0 upcalls registered for foreign access
        4 native libraries: dl, pthread, rt, z
[3/8] Building universe...                                                                               (5.4s @ 0.83GB)
[4/8] Parsing methods...      [**]                                                                       (4.1s @ 0.90GB)
[5/8] Inlining methods...     [***]                                                                      (2.5s @ 0.98GB)
[6/8] Compiling methods...    [****]                                                                    (20.7s @ 0.72GB)
[7/8] Laying out methods...   [**]                                                                       (2.9s @ 0.91GB)
[8/8] Creating image...       [**]                                                                       (3.5s @ 1.04GB)
  10.18MB (35.78%) for code area:    25,022 compilation units
  14.16MB (49.74%) for image heap:  221,395 objects and 151 resources
   4.12MB (14.48%) for other data
  28.46MB in total image size, 26.22MB in total file size
------------------------------------------------------------------------------------------------------------------------
Top 10 origins of code area:                                Top 10 object types in image heap:
   4.63MB java.base                                            3.01MB byte[] for code metadata
   2.19MB svm.jar (Native Image)                               2.72MB byte[] for java.lang.String
 564.37kB assertj-core-3.22.0.jar                              1.43MB java.lang.String
 481.34kB java.xml                                             1.12MB java.lang.Class
 310.55kB netty-buffer-4.2.1.Final.jar                       925.60kB com.oracle.svm.core.hub.DynamicHubCompanion
 262.81kB netty-common-4.2.1.Final.jar                       709.00kB byte[] for general heap data
 198.32kB junit-jupiter-engine-5.9.2.jar                     450.91kB java.util.concurrent.ConcurrentHashMap$Node
 132.29kB netty-transport-4.2.1.Final.jar                    402.50kB heap alignment
 121.73kB junit-platform-launcher-1.9.2.jar                  381.82kB byte[] for reflection metadata
 103.55kB java.logging                                       301.18kB java.util.HashMap$Node
 795.55kB for 31 more packages                                 2.70MB for 2042 more object types
                            Use '--emit build-report' to create a report with more details.
------------------------------------------------------------------------------------------------------------------------
Security report:
 - Binary includes Java deserialization.
 - CycloneDX SBOM with 24 component(s) is embedded in binary (1.18kB). 1 type(s) could not be associated to a component.
 - Advanced obfuscation not enabled; enable with '-H:AdvancedObfuscation=""' (experimental support).
------------------------------------------------------------------------------------------------------------------------
Recommendations:
 G1GC: Use the G1 GC ('--gc=G1') for improved latency and throughput.
 PGO:  Use Profile-Guided Optimizations ('--pgo') for improved throughput.
 FUTR: Use '--future-defaults=all' to prepare for future releases.
 HEAP: Set max heap for improved and more predictable memory usage.
 CPU:  Enable more CPU features with '-march=native' for improved performance.
------------------------------------------------------------------------------------------------------------------------
                        5.4s (6.3% of total time) in 742 GCs | Peak RSS: 1.85GB | CPU load: 3.63
------------------------------------------------------------------------------------------------------------------------
Build artifacts:
 /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/native/nativeTestCompile/io.netty.netty-codec-memcache_tests-tests (executable)
 /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/native/nativeTestCompile/libmanagement_ext.so (jdk_library)
 /tmp/native-image-build-output.json (build_info)
========================================================================================================================
Finished generating 'io.netty.netty-codec-memcache_tests-tests' in 1m 25s.
[native-image-plugin] Native Image written to: /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/native/nativeTestCompile

> Task :nativeTest FAILED
[junit-platform-native] WARNING: -djunit.platform.listeners.uid.tracking.output.dir not specified, trying to find test-ids on default Gradle/Maven locations. As this is a fallback mode, it could take a while. This should only happen if you are running tests executable manually and you didn't pass uid output directory with -djunit.platform.listeners.uid.tracking.output.dir=<path>.
[junit-platform-native] WARNING: Using test-ids from default Gradle project location:/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/test-results/test/testlist
[junit-platform-native] Running in 'test listener' mode using files matching pattern [junit-platform-unique-ids*] found in folder [/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/test-results/test/testlist] and its subfolders.
JUnit Platform on Native Image - report
----------------------------------------

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > serverCodecDecodesRequestsAndEncodesResponses() FAILED

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > decodesFragmentedBinaryResponseIntoMessageAndContentChunks() SUCCESSFUL

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > clientCodecEncodesRequestsAndDecodesResponses() FAILED

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > aggregatesDecodedBinaryResponseIntoFullMessage() SUCCESSFUL

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > aggregatesFragmentedSetRequestIntoFullMessage() SUCCESSFUL

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > clientCodecReportsMissingFullResponseWhenChannelCloses() FAILED

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > contentAndFullMessageHoldersCopyDuplicateRetainAndReplaceBuffers() SUCCESSFUL

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > encodesFullBinarySetRequestHeaderKeyExtrasAndContent() FAILED


Failures (4):
  JUnit Jupiter:Netty_codec_memcacheTest:serverCodecDecodesRequestsAndEncodesResponses()
    MethodSource [className = 'io_netty.netty_codec_memcache.Netty_codec_memcacheTest', methodName = 'serverCodecDecodesRequestsAndEncodesResponses', methodParameterTypes = '']
    => io.netty.handler.codec.EncoderException: java.lang.NullPointerException
       io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
       io.netty.channel.CombinedChannelDuplexHandler.write(CombinedChannelDuplexHandler.java:346)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:828)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:749)
       io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1031)
       [...]
     Caused by: java.lang.NullPointerException
       io.netty.util.internal.PlatformDependent0.safeConstructPutInt(PlatformDependent0.java:699)
       io.netty.util.internal.PlatformDependent.safeConstructPutInt(PlatformDependent.java:516)
       io.netty.util.internal.ReferenceCountUpdater.setInitialValue(ReferenceCountUpdater.java:67)
       io.netty.buffer.AdaptivePoolingAllocator$Chunk.<init>(AdaptivePoolingAllocator.java:840)
       io.netty.buffer.AdaptivePoolingAllocator$Magazine.newChunkAllocation(AdaptivePoolingAllocator.java:761)
       [...]
  JUnit Jupiter:Netty_codec_memcacheTest:clientCodecEncodesRequestsAndDecodesResponses()
    MethodSource [className = 'io_netty.netty_codec_memcache.Netty_codec_memcacheTest', methodName = 'clientCodecEncodesRequestsAndDecodesResponses', methodParameterTypes = '']
    => io.netty.handler.codec.EncoderException: java.lang.NullPointerException
       io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
       io.netty.channel.CombinedChannelDuplexHandler.write(CombinedChannelDuplexHandler.java:346)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:828)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:749)
       io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1031)
       [...]
     Caused by: java.lang.NullPointerException
       io.netty.util.internal.PlatformDependent0.safeConstructPutInt(PlatformDependent0.java:699)
       io.netty.util.internal.PlatformDependent.safeConstructPutInt(PlatformDependent.java:516)
       io.netty.util.internal.ReferenceCountUpdater.setInitialValue(ReferenceCountUpdater.java:67)
       io.netty.buffer.AdaptivePoolingAllocator$Chunk.<init>(AdaptivePoolingAllocator.java:840)
       io.netty.buffer.AdaptivePoolingAllocator$Magazine.newChunkAllocation(AdaptivePoolingAllocator.java:761)
       [...]
  JUnit Jupiter:Netty_codec_memcacheTest:clientCodecReportsMissingFullResponseWhenChannelCloses()
    MethodSource [className = 'io_netty.netty_codec_memcache.Netty_codec_memcacheTest', methodName = 'clientCodecReportsMissingFullResponseWhenChannelCloses', methodParameterTypes = '']
    => io.netty.handler.codec.EncoderException: java.lang.NullPointerException
       io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
       io.netty.channel.CombinedChannelDuplexHandler.write(CombinedChannelDuplexHandler.java:346)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:828)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:749)
       io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1031)
       [...]
     Caused by: java.lang.NullPointerException
       io.netty.util.internal.PlatformDependent0.safeConstructPutInt(PlatformDependent0.java:699)
       io.netty.util.internal.PlatformDependent.safeConstructPutInt(PlatformDependent.java:516)
       io.netty.util.internal.ReferenceCountUpdater.setInitialValue(ReferenceCountUpdater.java:67)
       io.netty.buffer.AdaptivePoolingAllocator$Chunk.<init>(AdaptivePoolingAllocator.java:840)
       io.netty.buffer.AdaptivePoolingAllocator$Magazine.newChunkAllocation(AdaptivePoolingAllocator.java:761)
       [...]
  JUnit Jupiter:Netty_codec_memcacheTest:encodesFullBinarySetRequestHeaderKeyExtrasAndContent()
    MethodSource [className = 'io_netty.netty_codec_memcache.Netty_codec_memcacheTest', methodName = 'encodesFullBinarySetRequestHeaderKeyExtrasAndContent', methodParameterTypes = '']
    => io.netty.handler.codec.EncoderException: java.lang.NullPointerException
       io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:830)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:749)
       io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1031)
       io.netty.channel.Channel.write(Channel.java:272)
       [...]
     Caused by: java.lang.NullPointerException
       io.netty.util.internal.PlatformDependent0.safeConstructPutInt(PlatformDependent0.java:699)
       io.netty.util.internal.PlatformDependent.safeConstructPutInt(PlatformDependent.java:516)
       io.netty.util.internal.ReferenceCountUpdater.setInitialValue(ReferenceCountUpdater.java:67)
       io.netty.buffer.AdaptivePoolingAllocator$Chunk.<init>(AdaptivePoolingAllocator.java:840)
       io.netty.buffer.AdaptivePoolingAllocator$Magazine.newChunkAllocation(AdaptivePoolingAllocator.java:761)
       [...]

Test run finished after 6 ms
[         3 containers found      ]
[         0 containers skipped    ]
[         3 containers started    ]
[         0 containers aborted    ]
[         3 containers successful ]
[         0 containers failed     ]
[         8 tests found           ]
[         0 tests skipped         ]
[         8 tests started         ]
[         0 tests aborted         ]
[         4 tests successful      ]
[         4 tests failed          ]


[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
17 actionable tasks: 7 executed, 10 up-to-date

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':nativeTest'.
> Process 'command '/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/native/nativeTestCompile/io.netty.netty-codec-memcache_tests-tests'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

BUILD FAILED in 1m 29s

> Task :tckTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':tckTest'.
> Test for io.netty:netty-codec-memcache:4.1.79.Final failed with exit code 1.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 1m 30s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'native-image run' for 4.2.1.Final at 2026-05-04T01:51:24Z with exit code 1 after 91s
❌ ERROR: 'native-image run' for 4.2.1.Final failed with exit code 1.
FAILED[native-image run][4.2.1.Final][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.2.1.Final" ./gradlew clean test -Pcoordinates="io.netty:netty-codec-memcache:4.1.79.Final"]
```

## native-image run on ubuntu-latest / JDK latest-ea / mode current-defaults

Failure kind: native-image run
OS: ubuntu-latest
JDK: latest-ea
Native-image mode: current-defaults
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.2.1.Final" ./gradlew clean test -Pcoordinates="io.netty:netty-codec-memcache:4.1.79.Final"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25296069081/job/74155118865
Last 300 lines of the log (excerpt truncated to fit the GitHub issue size limit):
```console
ux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/library-support.jar!/META-INF/native-image/com.oracle.svm/polyglot/native-image.properties+api=com.oracle.svm.polyglot.groovy.GroovyIndyInterfaceFeature' \
'-H:AddExports@jar:file:///opt/hostedtoolcache/graalvm-jdk-latest-ea_linux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/svm-guest.jar!/META-INF/native-image/org.graalvm.nativeimage.guest/native-image.properties+api=java.base/jdk.internal.misc=org.graalvm.nativeimage.guest' \
'-H:Features@jar:file:///opt/hostedtoolcache/graalvm-jdk-latest-ea_linux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/svm-guest.jar!/META-INF/native-image/org.graalvm.nativeimage.guest/native-image.properties+api=com.oracle.svm.guest.features.jdk.hosted.JavaNetHttpFeature' \
'-H:ClassInitialization@jar:file:///opt/hostedtoolcache/graalvm-jdk-latest-ea_linux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/svm-guest.jar!/META-INF/native-image/org.graalvm.nativeimage.guest/native-image.properties+api=sun.net.httpserver.simpleserver:run_time' \
'-H:ReachabilityMetadataResources@jar:file:///opt/hostedtoolcache/graalvm-jdk-latest-ea_linux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/svm-guest.jar!/META-INF/native-image/org.graalvm.nativeimage.guest/reachability-metadata.json+api=META-INF/native-image/org.graalvm.nativeimage.guest/reachability-metadata.json' \
-H:DriverWarningsCount=4
]
Warning: Option 'FallbackThreshold' is deprecated and might be removed in a future release: It no longer has any effect, and no replacement is available. Please refer to the GraalVM release notes.
========================================================================================================================
GraalVM Native Image: Generating 'io.netty.netty-codec-memcache_tests-tests' (executable)...
========================================================================================================================
For detailed information and explanations on the build output, visit:
https://github.com/oracle/graal/blob/master/docs/reference-manual/native-image/BuildOutput.md
------------------------------------------------------------------------------------------------------------------------
[1/8] Initializing...                                                                                   (6.6s @ 0.21GiB)
 Builder configuration:
 - Java version: 25.0.2+10-LTS, vendor version: Oracle GraalVM 25.1.0-dev+10.1
 - Graal compiler: optimization level: b, target machine: x86-64-v3, PGO: off
 - C compiler: gcc (linux, x86_64, 13.3.0)
 - Assertions: disabled, system assertions: disabled
 - 1 user-specific feature(s):
   - org.graalvm.junit.platform.JUnitPlatformFeature
 Image configuration:
 - Garbage collector: Serial GC (max heap size: 80% of RAM)
 - Assertions: disabled (class-specific config may apply), system assertions: disabled
------------------------------------------------------------------------------------------------------------------------
 1 experimental option(s) unlocked:
 - '-H:BuildOutputJSONFile' (origin(s): command line)
------------------------------------------------------------------------------------------------------------------------
Build resources:
 - 13.27GiB of memory (85% of system memory because $CI set to 'true')
 - 4 thread(s) (100.0% of 4 available processor(s), determined at start)
[junit-platform-native] Running in 'test listener' mode using files matching pattern [junit-platform-unique-ids*] found in folder [/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/test-results/test/testlist] and its subfolders.
[2/8] Performing analysis...  [******]                                                                 (31.7s @ 0.87GiB)
    8,697 types,   9,873 fields, and  39,324 methods found reachable
      369 types,     121 fields, and     778 methods registered for reflection
       60 types,      68 fields, and      52 methods registered for JNI access
      151 resource accesses registered with 262.39KiB total size
        4 native libraries: dl, pthread, rt, z
[3/8] Building universe...                                                                              (3.8s @ 0.97GiB)
[4/8] Parsing methods...      [**]                                                                      (3.9s @ 0.62GiB)
[5/8] Inlining methods...     [***]                                                                     (1.7s @ 0.66GiB)
[6/8] Compiling methods...    [****]                                                                   (17.6s @ 0.72GiB)
[7/8] Laying out methods...   [**]                                                                      (3.1s @ 0.87GiB)
[8/8] Creating image...       [**]                                                                      (3.3s @ 1.02GiB)
   8.79MiB (33.69%) for code area:    24,541 compilation units
  13.38MiB (51.24%) for image heap:  221,673 objects and 21 resources
   3.93MiB (15.07%) for other data
  26.10MiB in total image size, 24.00MiB in total file size
------------------------------------------------------------------------------------------------------------------------
Top 10 origins of code area:                                Top 10 object types in image heap:
 992.56KiB o.g.n.runtime/com.oracle.svm.core.code              2.75MiB byte[] for code metadata
 700.59KiB java.base/java.util                                 2.62MiB byte[] for string data
 408.45KiB java.base/jdk.internal.classfile.impl               1.40MiB java.base/java.lang.String
 305.26KiB java.base/java.lang                                 1.05MiB java.base/java.lang.Class
 287.57KiB io.netty.buffer                                   951.23KiB o.g.n.runtime/c.o.s.core.hub.DynamicHubCompanion
 228.82KiB java.base/java.text                               581.38KiB byte[] for general heap data
 226.63KiB java.base/java.util.concurrent                    519.79KiB byte[] for reflection metadata
 224.55KiB java.base/com.sun.crypto.provider                 337.11KiB o.g.n.runtime/c.o.s.c.h.ImageReflectionMetadata
 223.93KiB org.assertj.core.internal.bytebuddy.jar.asm       298.56KiB java.base/java.util.HashMap$Node
 164.05KiB java.base/java.lang.invoke                        262.79KiB byte[] for embedded resources
   5.12MiB for 325 more packages                               2.66MiB for 2040 more object types
                            Use '--emit build-report' to create a report with more details.
------------------------------------------------------------------------------------------------------------------------
Security report:
 - Binary includes Java deserialization.
 - CycloneDX SBOM with 23 component(s) is embedded in binary (1.18KiB). 1 type(s) could not be associated to a component.
 - Advanced obfuscation not enabled; enable with '-H:AdvancedObfuscation=""' (experimental support).
------------------------------------------------------------------------------------------------------------------------
Recommendations:
 G1GC: Use the G1 GC ('--gc=G1') for improved latency and throughput.
 PGO:  Use Profile-Guided Optimizations ('--pgo') for improved throughput.
 FUTR: Use '--future-defaults=all' to prepare for future releases.
 HEAP: Set max heap for improved and more predictable memory usage.
 CPU:  Enable more CPU features with '-march=native' for improved performance.
------------------------------------------------------------------------------------------------------------------------
                       5.8s (7.7% of total time) in 709 GCs | Peak RSS: 1.64GiB | CPU load: 3.59
------------------------------------------------------------------------------------------------------------------------
Build artifacts:
 /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/native/nativeTestCompile/io.netty.netty-codec-memcache_tests-tests (executable, 24.00MiB)
 /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/native/nativeTestCompile/libmanagement_ext.so (jdk_library, 32.23KiB)
 /tmp/native-image-build-output.json (build_info, 1.04KiB)
========================================================================================================================
Finished generating 'io.netty.netty-codec-memcache_tests-tests' in 1m 13s.

The build process encountered 5 warnings.

[native-image-plugin] Native Image written to: /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/native/nativeTestCompile

> Task :nativeTest FAILED
[junit-platform-native] WARNING: -djunit.platform.listeners.uid.tracking.output.dir not specified, trying to find test-ids on default Gradle/Maven locations. As this is a fallback mode, it could take a while. This should only happen if you are running tests executable manually and you didn't pass uid output directory with -djunit.platform.listeners.uid.tracking.output.dir=<path>.
[junit-platform-native] WARNING: Using test-ids from default Gradle project location:/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/test-results/test/testlist
[junit-platform-native] Running in 'test listener' mode using files matching pattern [junit-platform-unique-ids*] found in folder [/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/test-results/test/testlist] and its subfolders.
JUnit Platform on Native Image - report
----------------------------------------

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > serverCodecDecodesRequestsAndEncodesResponses() FAILED

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > decodesFragmentedBinaryResponseIntoMessageAndContentChunks() SUCCESSFUL

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > clientCodecEncodesRequestsAndDecodesResponses() FAILED

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > aggregatesDecodedBinaryResponseIntoFullMessage() SUCCESSFUL

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > aggregatesFragmentedSetRequestIntoFullMessage() SUCCESSFUL

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > clientCodecReportsMissingFullResponseWhenChannelCloses() FAILED

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > contentAndFullMessageHoldersCopyDuplicateRetainAndReplaceBuffers() SUCCESSFUL

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > encodesFullBinarySetRequestHeaderKeyExtrasAndContent() FAILED


Failures (4):
  JUnit Jupiter:Netty_codec_memcacheTest:serverCodecDecodesRequestsAndEncodesResponses()
    MethodSource [className = 'io_netty.netty_codec_memcache.Netty_codec_memcacheTest', methodName = 'serverCodecDecodesRequestsAndEncodesResponses', methodParameterTypes = '']
    => io.netty.handler.codec.EncoderException: java.lang.NullPointerException
       io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
       io.netty.channel.CombinedChannelDuplexHandler.write(CombinedChannelDuplexHandler.java:346)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:828)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:749)
       io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1031)
       [...]
     Caused by: java.lang.NullPointerException
       io.netty.util.internal.PlatformDependent0.safeConstructPutInt(PlatformDependent0.java:699)
       io.netty.util.internal.PlatformDependent.safeConstructPutInt(PlatformDependent.java:516)
       io.netty.util.internal.ReferenceCountUpdater.setInitialValue(ReferenceCountUpdater.java:67)
       io.netty.buffer.AdaptivePoolingAllocator$Chunk.<init>(AdaptivePoolingAllocator.java:840)
       io.netty.buffer.AdaptivePoolingAllocator$Magazine.newChunkAllocation(AdaptivePoolingAllocator.java:761)
       [...]
  JUnit Jupiter:Netty_codec_memcacheTest:clientCodecEncodesRequestsAndDecodesResponses()
    MethodSource [className = 'io_netty.netty_codec_memcache.Netty_codec_memcacheTest', methodName = 'clientCodecEncodesRequestsAndDecodesResponses', methodParameterTypes = '']
    => io.netty.handler.codec.EncoderException: java.lang.NullPointerException
       io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
       io.netty.channel.CombinedChannelDuplexHandler.write(CombinedChannelDuplexHandler.java:346)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:828)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:749)
       io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1031)
       [...]
     Caused by: java.lang.NullPointerException
       io.netty.util.internal.PlatformDependent0.safeConstructPutInt(PlatformDependent0.java:699)
       io.netty.util.internal.PlatformDependent.safeConstructPutInt(PlatformDependent.java:516)
       io.netty.util.internal.ReferenceCountUpdater.setInitialValue(ReferenceCountUpdater.java:67)
       io.netty.buffer.AdaptivePoolingAllocator$Chunk.<init>(AdaptivePoolingAllocator.java:840)
       io.netty.buffer.AdaptivePoolingAllocator$Magazine.newChunkAllocation(AdaptivePoolingAllocator.java:761)
       [...]
  JUnit Jupiter:Netty_codec_memcacheTest:clientCodecReportsMissingFullResponseWhenChannelCloses()
    MethodSource [className = 'io_netty.netty_codec_memcache.Netty_codec_memcacheTest', methodName = 'clientCodecReportsMissingFullResponseWhenChannelCloses', methodParameterTypes = '']
    => io.netty.handler.codec.EncoderException: java.lang.NullPointerException
       io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
       io.netty.channel.CombinedChannelDuplexHandler.write(CombinedChannelDuplexHandler.java:346)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:828)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:749)
       io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1031)
       [...]
     Caused by: java.lang.NullPointerException
       io.netty.util.internal.PlatformDependent0.safeConstructPutInt(PlatformDependent0.java:699)
       io.netty.util.internal.PlatformDependent.safeConstructPutInt(PlatformDependent.java:516)
       io.netty.util.internal.ReferenceCountUpdater.setInitialValue(ReferenceCountUpdater.java:67)
       io.netty.buffer.AdaptivePoolingAllocator$Chunk.<init>(AdaptivePoolingAllocator.java:840)
       io.netty.buffer.AdaptivePoolingAllocator$Magazine.newChunkAllocation(AdaptivePoolingAllocator.java:761)
       [...]
  JUnit Jupiter:Netty_codec_memcacheTest:encodesFullBinarySetRequestHeaderKeyExtrasAndContent()
    MethodSource [className = 'io_netty.netty_codec_memcache.Netty_codec_memcacheTest', methodName = 'encodesFullBinarySetRequestHeaderKeyExtrasAndContent', methodParameterTypes = '']
    => io.netty.handler.codec.EncoderException: java.lang.NullPointerException
       io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:830)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:749)
       io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1031)
       io.netty.channel.Channel.write(Channel.java:272)
       [...]
     Caused by: java.lang.NullPointerException
       io.netty.util.internal.PlatformDependent0.safeConstructPutInt(PlatformDependent0.java:699)
       io.netty.util.internal.PlatformDependent.safeConstructPutInt(PlatformDependent.java:516)
       io.netty.util.internal.ReferenceCountUpdater.setInitialValue(ReferenceCountUpdater.java:67)
       io.netty.buffer.AdaptivePoolingAllocator$Chunk.<init>(AdaptivePoolingAllocator.java:840)
       io.netty.buffer.AdaptivePoolingAllocator$Magazine.newChunkAllocation(AdaptivePoolingAllocator.java:761)
       [...]

Test run finished after 5 ms
[         3 containers found      ]
[         0 containers skipped    ]
[         3 containers started    ]
[         0 containers aborted    ]
[         3 containers successful ]
[         0 containers failed     ]
[         8 tests found           ]
[         0 tests skipped         ]
[         8 tests started         ]
[         0 tests aborted         ]
[         4 tests successful      ]
[         4 tests failed          ]


[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
17 actionable tasks: 7 executed, 10 up-to-date

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':nativeTest'.
> Process 'command '/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/native/nativeTestCompile/io.netty.netty-codec-memcache_tests-tests'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

BUILD FAILED in 1m 17s

> Task :tckTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':tckTest'.
> Test for io.netty:netty-codec-memcache:4.1.79.Final failed with exit code 1.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 1m 19s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'native-image run' for 4.2.1.Final at 2026-05-04T01:46:35Z with exit code 1 after 80s
❌ ERROR: 'native-image run' for 4.2.1.Final failed with exit code 1.
FAILED[native-image run][4.2.1.Final][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.2.1.Final" ./gradlew clean test -Pcoordinates="io.netty:netty-codec-memcache:4.1.79.Final"]
```

## native-image run on ubuntu-latest / JDK latest-ea / mode future-defaults-all

Failure kind: native-image run
OS: ubuntu-latest
JDK: latest-ea
Native-image mode: future-defaults-all
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="4.2.1.Final" ./gradlew clean test -Pcoordinates="io.netty:netty-codec-memcache:4.1.79.Final"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25296069081/job/74155118864
Last 300 lines of the log (excerpt truncated to fit the GitHub issue size limit):
```console
ile:///opt/hostedtoolcache/graalvm-jdk-latest-ea_linux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/library-support.jar!/META-INF/native-image/com.oracle.svm/polyglot/native-image.properties+api=com.oracle.svm.polyglot.groovy.GroovyIndyInterfaceFeature' \
'-H:AddExports@jar:file:///opt/hostedtoolcache/graalvm-jdk-latest-ea_linux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/svm-guest.jar!/META-INF/native-image/org.graalvm.nativeimage.guest/native-image.properties+api=java.base/jdk.internal.misc=org.graalvm.nativeimage.guest' \
'-H:Features@jar:file:///opt/hostedtoolcache/graalvm-jdk-latest-ea_linux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/svm-guest.jar!/META-INF/native-image/org.graalvm.nativeimage.guest/native-image.properties+api=com.oracle.svm.guest.features.jdk.hosted.JavaNetHttpFeature' \
'-H:ClassInitialization@jar:file:///opt/hostedtoolcache/graalvm-jdk-latest-ea_linux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/svm-guest.jar!/META-INF/native-image/org.graalvm.nativeimage.guest/native-image.properties+api=sun.net.httpserver.simpleserver:run_time' \
'-H:ReachabilityMetadataResources@jar:file:///opt/hostedtoolcache/graalvm-jdk-latest-ea_linux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/svm-guest.jar!/META-INF/native-image/org.graalvm.nativeimage.guest/reachability-metadata.json+api=META-INF/native-image/org.graalvm.nativeimage.guest/reachability-metadata.json' \
-H:DriverWarningsCount=4
]
Warning: Option 'FallbackThreshold' is deprecated and might be removed in a future release: It no longer has any effect, and no replacement is available. Please refer to the GraalVM release notes.
========================================================================================================================
GraalVM Native Image: Generating 'io.netty.netty-codec-memcache_tests-tests' (executable)...
========================================================================================================================
For detailed information and explanations on the build output, visit:
https://github.com/oracle/graal/blob/master/docs/reference-manual/native-image/BuildOutput.md
------------------------------------------------------------------------------------------------------------------------
[1/8] Initializing...                                                                                   (6.5s @ 0.21GiB)
 Builder configuration:
 - Java version: 25.0.2+10-LTS, vendor version: Oracle GraalVM 25.1.0-dev+10.1
 - Graal compiler: optimization level: b, target machine: x86-64-v3, PGO: off
 - C compiler: gcc (linux, x86_64, 13.3.0)
 - Assertions: disabled, system assertions: disabled
 - 1 user-specific feature(s):
   - org.graalvm.junit.platform.JUnitPlatformFeature
 Image configuration:
 - Garbage collector: Serial GC (max heap size: 80% of RAM)
 - Assertions: disabled (class-specific config may apply), system assertions: disabled
------------------------------------------------------------------------------------------------------------------------
 1 experimental option(s) unlocked:
 - '-H:BuildOutputJSONFile' (origin(s): command line)
------------------------------------------------------------------------------------------------------------------------
Build resources:
 - 13.27GiB of memory (85% of system memory because $CI set to 'true')
 - 4 thread(s) (100.0% of 4 available processor(s), determined at start)
[junit-platform-native] Running in 'test listener' mode using files matching pattern [junit-platform-unique-ids*] found in folder [/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/test-results/test/testlist] and its subfolders.
[2/8] Performing analysis...  [*****]                                                                  (34.1s @ 0.94GiB)
    9,237 types,  10,675 fields, and  42,353 methods found reachable
      497 types,     121 fields, and     982 methods registered for reflection
       60 types,      68 fields, and      52 methods registered for JNI access
      153 resource accesses registered with 262.53KiB total size
        4 native libraries: dl, pthread, rt, z
[3/8] Building universe...                                                                              (4.2s @ 1.05GiB)
[4/8] Parsing methods...      [**]                                                                      (4.3s @ 0.78GiB)
[5/8] Inlining methods...     [***]                                                                     (1.7s @ 0.83GiB)
[6/8] Compiling methods...    [****]                                                                   (19.5s @ 0.85GiB)
[7/8] Laying out methods...   [**]                                                                      (3.3s @ 1.04GiB)
[8/8] Creating image...       [**]                                                                      (3.8s @ 1.18GiB)
   9.91MiB (33.22%) for code area:    26,849 compilation units
  15.75MiB (52.78%) for image heap:  260,606 objects and 23 resources
   4.18MiB (14.00%) for other data
  29.84MiB in total image size, 27.56MiB in total file size
------------------------------------------------------------------------------------------------------------------------
Top 10 origins of code area:                                Top 10 object types in image heap:
   1.10MiB o.g.n.runtime/com.oracle.svm.core.code              3.11MiB byte[] for code metadata
 707.56KiB java.base/java.util                                 2.91MiB byte[] for string data
 408.42KiB java.base/jdk.internal.classfile.impl               1.54MiB java.base/java.lang.String
 314.82KiB java.base/java.lang                                 1.12MiB java.base/java.lang.Class
 287.57KiB io.netty.buffer                                     0.99MiB o.g.n.runtime/c.o.s.core.hub.DynamicHubCompanion
 267.01KiB java.base/com.sun.crypto.provider                 929.93KiB byte[] for general heap data
 231.38KiB java.base/java.text                               587.83KiB byte[] for reflection metadata
 226.75KiB java.base/java.util.concurrent                    453.27KiB heap alignment
 223.93KiB org.assertj.core.internal.bytebuddy.jar.asm       362.25KiB java.base/java.lang.Object[]
 192.84KiB java.base/sun.security.provider                   358.20KiB o.g.n.runtime/c.o.s.c.h.ImageReflectionMetadata
   6.02MiB for 335 more packages                               3.45MiB for 2196 more object types
                            Use '--emit build-report' to create a report with more details.
------------------------------------------------------------------------------------------------------------------------
Security report:
 - Binary includes Java deserialization.
 - CycloneDX SBOM with 23 component(s) is embedded in binary (1.18KiB). 1 type(s) could not be associated to a component.
 - Advanced obfuscation not enabled; enable with '-H:AdvancedObfuscation=""' (experimental support).
------------------------------------------------------------------------------------------------------------------------
Recommendations:
 G1GC: Use the G1 GC ('--gc=G1') for improved latency and throughput.
 PGO:  Use Profile-Guided Optimizations ('--pgo') for improved throughput.
 HEAP: Set max heap for improved and more predictable memory usage.
 CPU:  Enable more CPU features with '-march=native' for improved performance.
------------------------------------------------------------------------------------------------------------------------
                       6.5s (8.1% of total time) in 809 GCs | Peak RSS: 1.81GiB | CPU load: 3.60
------------------------------------------------------------------------------------------------------------------------
Build artifacts:
 /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/native/nativeTestCompile/io.netty.netty-codec-memcache_tests-tests (executable, 27.56MiB)
 /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/native/nativeTestCompile/libmanagement_ext.so (jdk_library, 32.23KiB)
 /tmp/native-image-build-output.json (build_info, 1.04KiB)
========================================================================================================================
Finished generating 'io.netty.netty-codec-memcache_tests-tests' in 1m 19s.

The build process encountered 5 warnings.

[native-image-plugin] Native Image written to: /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/native/nativeTestCompile

> Task :nativeTest FAILED
[junit-platform-native] WARNING: -djunit.platform.listeners.uid.tracking.output.dir not specified, trying to find test-ids on default Gradle/Maven locations. As this is a fallback mode, it could take a while. This should only happen if you are running tests executable manually and you didn't pass uid output directory with -djunit.platform.listeners.uid.tracking.output.dir=<path>.
[junit-platform-native] WARNING: Using test-ids from default Gradle project location:/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/test-results/test/testlist
[junit-platform-native] Running in 'test listener' mode using files matching pattern [junit-platform-unique-ids*] found in folder [/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/test-results/test/testlist] and its subfolders.
JUnit Platform on Native Image - report
----------------------------------------

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > serverCodecDecodesRequestsAndEncodesResponses() FAILED

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > decodesFragmentedBinaryResponseIntoMessageAndContentChunks() SUCCESSFUL

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > clientCodecEncodesRequestsAndDecodesResponses() FAILED

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > aggregatesDecodedBinaryResponseIntoFullMessage() SUCCESSFUL

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > aggregatesFragmentedSetRequestIntoFullMessage() SUCCESSFUL

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > clientCodecReportsMissingFullResponseWhenChannelCloses() FAILED

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > contentAndFullMessageHoldersCopyDuplicateRetainAndReplaceBuffers() SUCCESSFUL

io_netty.netty_codec_memcache.Netty_codec_memcacheTest > encodesFullBinarySetRequestHeaderKeyExtrasAndContent() FAILED


Failures (4):
  JUnit Jupiter:Netty_codec_memcacheTest:serverCodecDecodesRequestsAndEncodesResponses()
    MethodSource [className = 'io_netty.netty_codec_memcache.Netty_codec_memcacheTest', methodName = 'serverCodecDecodesRequestsAndEncodesResponses', methodParameterTypes = '']
    => io.netty.handler.codec.EncoderException: java.lang.NullPointerException
       io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
       io.netty.channel.CombinedChannelDuplexHandler.write(CombinedChannelDuplexHandler.java:346)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:828)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:749)
       io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1031)
       [...]
     Caused by: java.lang.NullPointerException
       io.netty.util.internal.PlatformDependent0.safeConstructPutInt(PlatformDependent0.java:699)
       io.netty.util.internal.PlatformDependent.safeConstructPutInt(PlatformDependent.java:516)
       io.netty.util.internal.ReferenceCountUpdater.setInitialValue(ReferenceCountUpdater.java:67)
       io.netty.buffer.AdaptivePoolingAllocator$Chunk.<init>(AdaptivePoolingAllocator.java:840)
       io.netty.buffer.AdaptivePoolingAllocator$Magazine.newChunkAllocation(AdaptivePoolingAllocator.java:761)
       [...]
  JUnit Jupiter:Netty_codec_memcacheTest:clientCodecEncodesRequestsAndDecodesResponses()
    MethodSource [className = 'io_netty.netty_codec_memcache.Netty_codec_memcacheTest', methodName = 'clientCodecEncodesRequestsAndDecodesResponses', methodParameterTypes = '']
    => io.netty.handler.codec.EncoderException: java.lang.NullPointerException
       io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
       io.netty.channel.CombinedChannelDuplexHandler.write(CombinedChannelDuplexHandler.java:346)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:828)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:749)
       io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1031)
       [...]
     Caused by: java.lang.NullPointerException
       io.netty.util.internal.PlatformDependent0.safeConstructPutInt(PlatformDependent0.java:699)
       io.netty.util.internal.PlatformDependent.safeConstructPutInt(PlatformDependent.java:516)
       io.netty.util.internal.ReferenceCountUpdater.setInitialValue(ReferenceCountUpdater.java:67)
       io.netty.buffer.AdaptivePoolingAllocator$Chunk.<init>(AdaptivePoolingAllocator.java:840)
       io.netty.buffer.AdaptivePoolingAllocator$Magazine.newChunkAllocation(AdaptivePoolingAllocator.java:761)
       [...]
  JUnit Jupiter:Netty_codec_memcacheTest:clientCodecReportsMissingFullResponseWhenChannelCloses()
    MethodSource [className = 'io_netty.netty_codec_memcache.Netty_codec_memcacheTest', methodName = 'clientCodecReportsMissingFullResponseWhenChannelCloses', methodParameterTypes = '']
    => io.netty.handler.codec.EncoderException: java.lang.NullPointerException
       io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
       io.netty.channel.CombinedChannelDuplexHandler.write(CombinedChannelDuplexHandler.java:346)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:828)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:749)
       io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1031)
       [...]
     Caused by: java.lang.NullPointerException
       io.netty.util.internal.PlatformDependent0.safeConstructPutInt(PlatformDependent0.java:699)
       io.netty.util.internal.PlatformDependent.safeConstructPutInt(PlatformDependent.java:516)
       io.netty.util.internal.ReferenceCountUpdater.setInitialValue(ReferenceCountUpdater.java:67)
       io.netty.buffer.AdaptivePoolingAllocator$Chunk.<init>(AdaptivePoolingAllocator.java:840)
       io.netty.buffer.AdaptivePoolingAllocator$Magazine.newChunkAllocation(AdaptivePoolingAllocator.java:761)
       [...]
  JUnit Jupiter:Netty_codec_memcacheTest:encodesFullBinarySetRequestHeaderKeyExtrasAndContent()
    MethodSource [className = 'io_netty.netty_codec_memcache.Netty_codec_memcacheTest', methodName = 'encodesFullBinarySetRequestHeaderKeyExtrasAndContent', methodParameterTypes = '']
    => io.netty.handler.codec.EncoderException: java.lang.NullPointerException
       io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:830)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:749)
       io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1031)
       io.netty.channel.Channel.write(Channel.java:272)
       [...]
     Caused by: java.lang.NullPointerException
       io.netty.util.internal.PlatformDependent0.safeConstructPutInt(PlatformDependent0.java:699)
       io.netty.util.internal.PlatformDependent.safeConstructPutInt(PlatformDependent.java:516)
       io.netty.util.internal.ReferenceCountUpdater.setInitialValue(ReferenceCountUpdater.java:67)
       io.netty.buffer.AdaptivePoolingAllocator$Chunk.<init>(AdaptivePoolingAllocator.java:840)
       io.netty.buffer.AdaptivePoolingAllocator$Magazine.newChunkAllocation(AdaptivePoolingAllocator.java:761)
       [...]

Test run finished after 5 ms
[         3 containers found      ]
[         0 containers skipped    ]
[         3 containers started    ]
[         0 containers aborted    ]
[         3 containers successful ]
[         0 containers failed     ]
[         8 tests found           ]
[         0 tests skipped         ]
[         8 tests started         ]
[         0 tests aborted         ]
[         4 tests successful      ]
[         4 tests failed          ]


[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':nativeTest'.
> Process 'command '/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build/native/nativeTestCompile/io.netty.netty-codec-memcache_tests-tests'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

BUILD FAILED in 1m 23s

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
17 actionable tasks: 7 executed, 10 up-to-date

> Task :tckTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':tckTest'.
> Test for io.netty:netty-codec-memcache:4.1.79.Final failed with exit code 1.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 1m 24s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'native-image run' for 4.2.1.Final at 2026-05-04T01:48:23Z with exit code 1 after 85s
❌ ERROR: 'native-image run' for 4.2.1.Final failed with exit code 1.
FAILED[native-image run][4.2.1.Final][GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="4.2.1.Final" ./gradlew clean test -Pcoordinates="io.netty:netty-codec-memcache:4.1.79.Final"]
```

--- a/metadata/io.netty/netty-codec-memcache/4.2.1.Final/reachability-metadata.json
+++ b/metadata/io.netty/netty-codec-memcache/4.2.1.Final/reachability-metadata.json
@@ -1,0 +1,55 @@
+{
+  "reflection": [
+    {
+      "condition": {
+        "typeReached": "io.netty.handler.codec.memcache.binary.BinaryMemcacheRequestDecoder"
+      },
+      "type": "io.netty.buffer.AbstractByteBufAllocator",
+      "allDeclaredMethods": true
+    },
+    {
+      "condition": {
+        "typeReached": "io.netty.handler.codec.memcache.binary.BinaryMemcacheRequestDecoder"
+      },
+      "type": "io.netty.util.AbstractReferenceCounted",
+      "fields": [
+        {
+          "name": "refCnt"
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "io.netty.handler.codec.memcache.binary.AbstractBinaryMemcacheEncoder"
+      },
+      "type": "io.netty.util.internal.shaded.org.jctools.queues.MpscArrayQueueConsumerIndexField",
+      "fields": [
+        {
+          "name": "consumerIndex"
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "io.netty.handler.codec.memcache.binary.AbstractBinaryMemcacheEncoder"
+      },
+      "type": "io.netty.util.internal.shaded.org.jctools.queues.MpscArrayQueueProducerIndexField",
+      "fields": [
+        {
+          "name": "producerIndex"
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "io.netty.handler.codec.memcache.binary.AbstractBinaryMemcacheEncoder"
+      },
+      "type": "io.netty.util.internal.shaded.org.jctools.queues.MpscArrayQueueProducerLimitField",
+      "fields": [
+        {
+          "name": "producerLimit"
+        }
+      ]
+    }
+  ]
+}

--- a/metadata/io.netty/netty-codec-memcache/index.json
+++ b/metadata/io.netty/netty-codec-memcache/index.json
@@ -1,6 +1,21 @@
 [
   {
     "latest" : true,
+    "metadata-version" : "4.2.1.Final",
+    "test-version" : "4.1.74.Final",
+    "source-code-url" : "https://repo.maven.apache.org/maven2/io/netty/netty-codec-memcache/$version$/netty-codec-memcache-$version$-sources.jar",
+    "repository-url" : "https://github.com/netty/netty",
+    "test-code-url" : "https://repo.maven.apache.org/maven2/io/netty/netty-codec-memcache/$version$/netty-codec-memcache-$version$-test-sources.jar",
+    "documentation-url" : "https://repo.maven.apache.org/maven2/io/netty/netty-codec-memcache/$version$/netty-codec-memcache-$version$-javadoc.jar",
+    "description" : "Netty Codec Memcache provides encoders, decoders, message types, and aggregation support for the memcached binary and text protocols. It lets Netty-based clients and servers parse and produce Memcache requests and responses in asynchronous channel pipelines.",
+    "tested-versions" : [
+      "4.2.1.Final"
+    ],
+    "allowed-packages" : [
+      "io.netty.handler.codec.memcache"
+    ]
+  },
+  {
     "metadata-version" : "4.1.79.Final",
     "test-version" : "4.1.74.Final",
     "source-code-url" : "https://repo.maven.apache.org/maven2/io/netty/netty-codec-memcache/$version$/netty-codec-memcache-$version$-sources.jar",

--- a/stats/io.netty/netty-codec-memcache/4.2.1.Final/stats.json
+++ b/stats/io.netty/netty-codec-memcache/4.2.1.Final/stats.json
@@ -1,0 +1,31 @@
+{
+  "versions" : [ {
+    "version" : "4.2.1.Final",
+    "dynamicAccess" : {
+      "breakdown" : { },
+      "coverageRatio" : 1.0,
+      "coveredCalls" : 0,
+      "totalCalls" : 0
+    },
+    "libraryCoverage" : {
+      "instruction" : {
+        "covered" : 1417,
+        "missed" : 552,
+        "ratio" : 0.719655,
+        "total" : 1969
+      },
+      "line" : {
+        "covered" : 372,
+        "missed" : 158,
+        "ratio" : 0.701887,
+        "total" : 530
+      },
+      "method" : {
+        "covered" : 109,
+        "missed" : 58,
+        "ratio" : 0.652695,
+        "total" : 167
+      }
+    }
+  } ]
+}

--- a/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build.gradle
+++ b/tests/src/io.netty/netty-codec-memcache/4.1.74.Final/build.gradle
@@ -15,6 +15,16 @@ dependencies {
     testImplementation 'org.assertj:assertj-core:3.22.0'
 }
 
+tasks.named("nativeTest") {
+    if (libraryVersion == '4.2.1.Final') {
+        // Netty 4.2.x requires explicit unsafe access on Java 25 native runs.
+        runtimeArgs.addAll([
+                '-Dsun.misc.unsafe.memory.access=allow',
+                '-Dio.netty.tryReflectionSetAccessible=true'
+        ])
+    }
+}
+
 graalvmNative {
     agent {
         defaultMode = "conditional"

